protocol/op-contracts-manager-single-release-redesign.md

```diff
@@ -0,0 +1,67 @@
+# Purpose
+
+The original OPContractsManager (OPCM) design, outlined [here](./op-contracts-manager-arch.md), defined a singleton proxied contract that deployed the L1 contracts for standard OP Stack chains in a single transaction.
+
+Implementing upgrade functionality for OP Stack chains via OPCM is now a key goal. As we approach this feature, we've identified several interim adjustments to OPCM that will that will not only enable future upgradability but also simplify its overall design. Therefore, the purpose of this design document is to streamline OPCM’s architecture, making it easier to build and integrate upgradability features.
+
+# Summary
+
+OPCM will transition from its current singleton architecture to a multi-deployment model, with each deployment focused on a specific L1 smart contract release version. To achieve this, the following changes need to be implemented:
+
+- Remove the [proxy](https://github.com/ethereum-optimism/optimism/blob/4c015e3a36f8910e2cf8b447d62ab4c44b944cca/packages/contracts-bedrock/scripts/deploy/DeployImplementations.s.sol#L545) from OPCM.
+- Remove [initialize](https://github.com/ethereum-optimism/optimism/blob/28283a927e3124fa0b2cf8d47d1a734e95478215/packages/contracts-bedrock/src/L1/OPContractsManager.sol#L210) functionality from OPCM and introduce standard constructor initialization.
+- Add versioning to OPCM to expose which L1 contracts are deployed for the specific OPCM deployment e.g. `string constant l1ContractsVersion = "op-contracts/v1.6.0";` or something more nuanced that shows the relationship between the version string and the [standard release implementation addresses](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-mainnet.toml#L9).
```

Comment:
I think `"op-contracts/v1.6.0"` is fine. Best to stay in alignment with the release process.

Reply:
I would also add that the [OPCM semver](https://github.com/ethereum-optimism/optimism/blob/feat/isthmus-contracts/packages/contracts-bedrock/src/L1/OPContractsManager.sol#L133-L134) should change to just match the release version, ie. we'd have: 

```
string public constant version = "1.0.0-beta.21";

function l1ContractsVersion() public returns(string) {
  return string.concat("op-contracts/", version);
}
```

Also consider `release()` for the function name.

Reply:
The versioning bullet point here is meant to be a separate version than the regular semver that exists on all out contracts. 

OPCM should continue to include the semver as normal. But additionally, it should include the `op-contracts/v*.*.*` L1 smart contract release that it deploys. The code you provided seems like it is mixing the versions. I was thinking something more along the lines of:

```solidity
/// @custom:semver 1.0.0-beta.21
string public constant version = "1.0.0-beta.21";

// @notice L1 smart contracts release deployed by this version of OPCM.
string constant l1ContractsVersion = "op-contracts/v1.6.0";
```

Reply:
I've clarified this in the following commit: https://github.com/ethereum-optimism/design-docs/pull/156/commits/2011fb5d2c4fa482ecee2127d59376465129142d

Reply:
Closing this based on conversation in discord: https://discord.com/channels/1244729134312198194/1273519241076670526/1303831875852304455

Reply:
I outlined the discussion here: https://github.com/ethereum-optimism/design-docs/pull/156/commits/7f50c4f7bd700be7d8664306acc31885d2cb71c3